### DISCUSSION TITLE
How to setup references to another loaded program via the API

### DISCUSSION THREAD

**Discussion by fmagin** [NationalSecurityAgency/ghidra#5936] [Q&A]
I vaguely remember that Ghidra had the option to associate programs with another program, to create direct links in the listing that allowed jumping to the second program if clicking on a reference inside the first program. This magically happened with IIRC ELF binaries if the library was import before the main binary. This also allowed viewing multiple disassembly listings (for the two programs) inside the same CodeBrowser.

The problem is that I deal with Mach-O nowadays where this isn't supported yet, and I'd like to at least write a small code snippet for myself that I can use in the REPL to manually setup references that I often need to follow between some main binary and its library.

What is the basic API function that allows me to declare that some location/address inside the main program contains a pointer to a symbol in a different program? I.e. the kind of location that by default contains a pointer to the `EXTERNAL` segment. If it exists, I'd also be fine with some UI Dialog if there is one to do this already, then I can also find the underlying implementation to automate this at scale

**Comment by ryanmkurtz** [comment]
Before I go into any code snippets, I'd like to inquire about what Mach-O limitations you are currently facing?  I've put a fair amount of effort into improving this and am curious if something is not working as I intended, or if I overlooked something that needs to happen.

In the GUI, the linkage can be managed with `Window -> External Programs`, which maps library name to an already-imported Ghidra program.  In the `Symbol Tree` under `Imports`, you can also right click on the libraries (if they exist) and then do `Set External Program`.

**Reply by ryanmkurtz** [reply]
I now recall that a lot of the improvements I have made have not been released yet...they are currently in the `master` branch scheduled for release in Ghidra 10.5.

**Reply by ryanmkurtz** [reply]
Here are the relevant features/fixes for 10.5:
* The MachoLoader now creates thunks on stubs.
* Importing libraries that are referenced by absolute path (such as with Mach-O) now get saved to the project with their folder structure intact.  This fixes a potential DuplicateKeyException that could occur when using a Recursive Library Load Depth greater than 1, and removes any ambiguity that could occur when linking a program to its libraries.
* The MachoLoader now uses binding information (if present) to associate libraries with imported symbol name without the need for those libraries to be already present/loaded in the project.
* Improved library import log messages.
* When loading System Libraries From Disk on macOS, the dyld_shared_cache will be searched for in more default locations.

**Reply by fmagin** [reply]
Oh, sweet, I didn't look into the new changes in 10.5 yet, will give them a try now.
I had resigned myself that Mach-O/ObjC is niche enough that I need to implement things myself.

> The MachoLoader now uses binding information (if present) to associate libraries with imported symbol name without the need for those libraries to be already present/loaded in the project.

That specifically was one of the big things I was missing. Not necessarily the thing I requested in this post, but an important step towards automatically handling it I think

**Reply by ryanmkurtz** [reply]
Cool, let me know if you hit any problems or need more capabilities that aren't there yet.

**Reply by ryanmkurtz** [reply]
Also, these other things might be of interest to you too, if you are digging into components within the dyld_shared_cache.
* The dyld_shared_cache filesystem can now extract files for stubs and standalone data.
* dyld_shared_cache components extracted from Ghidra's DyldCacheFileSystem can now be added together on-demand with the Add To Program feature.  Broken references can be automatically resolved by right clicking on them and clicking References -> Add To Program.

**Reply by fmagin** [reply]
Well, my dev setup on my laptop is in a sorry state, will try this tomorrow when I'm back at my main machine.

I'm still curious what the actual APIs is for referencing symbols from another program, and what the general logic/design around that is.
That has been one of the killer features of Ghidra for me ever since it was publicly released, that was an absolute pain with IDA.
And in the medium term I will be dealing with (dataflow) graphs in the graph viewer where some nodes belong to addresses inside the main program, and some to addresses inside a library, so I'm trying to understand how that will work.

**Comment by ryanmkurtz** [comment]
To get back to your main question about the API, check out [AssociateExternalPELibrariesScript.java](https://github.com/NationalSecurityAgency/ghidra/blob/af0ea1e7396d922e1cc0f604aa84af8244ba3fd0/Ghidra/Features/Base/ghidra_scripts/AssociateExternalPELibrariesScript.java).  You can see it's using an instance of `ExternalManager` to link in external programs.  This script is assuming that there are already "Name" entries in the `External Programs` window...it's just linking in the correct Ghidra Program.  With the 10.5 changes you've been testing, you hopefully have these Name entries and can do something similar.

**Reply by fmagin** [reply]
Yes, this mostly works now, found two issues with this still, will be opening proper bug reports in a few minutes

**Reply by fmagin** [reply]
Okay, I think I understand this part now, but this now raises some additional questions:

Usually for msgSend call on some selector "foobar", I determine the actual implementation being called (by figuring out the class the selector is being called on). 

I can then set this as the primary reference at this callsite, which changes the the function call in the decompiler to the actual function.

```kotlin
                val ref = program.referenceManager.addMemoryReference(
                    callsite, //FROM
                    func.entryPoint, //TO
                    RefType.UNCONDITIONAL_CALL,
                    SourceType.USER_DEFINED,
                    0)
                program.referenceManager.setPrimary(ref, true)
```

How would I do this elegantly if the function is from a different program? One way I could think of is to create an extra location in the `EXTERNAL` segment for this function inside the first program, add a reference to that external symbol at the callsite. I could then associate that symbol with the second program, and set the external location address for that symbol.

My problem is that I don't know the underlying design well enough to know if this actually works and if this breaks some assumption because I'm introducing fake symbols

The desired behavior overall would be that I click on the call inside the decompiler output and it jumps to the function in the second program, and the decompiler then shows the called function (just like it would be if it was all inside one program)

**Reply by ryanmkurtz** [reply]
As far as I know, we don't have the capability to jump directly to the external program from the callsite...this is a known limitation.  Our external locations are always put on stubs or some import address table entry, so it's always an extra double click.  

Creating entries in the EXTERNAL block entry is a strategy we use when a relocation needs to takes place that is outside of the current program.  We make the new entry in the EXTERNAL block (getting the size right can be an issue for data), and then we setup a new ExternalLocation on that new memory that points to the other program.  It's typically a thunk function.

**Reply by ryanmkurtz** [reply]
Before coding anything up, i recommend setting things up in the GUI the way you like them, and then recreating that in code.  That's an easier way to see what is possible.

**Reply by fmagin** [reply]
An extra double click seems like an acceptable limitation, as long as I don't have to search through the symbol tree manually.

> Our external locations are always put on stubs or some import address table entry, so it's always an extra double click.

How does that "putting an external location on something" work via the GUI?

Also an implementation of an ObjC method will have no stub or reference( like an  IAT entry) in binary that calls it, because it's only indirectly called via the msgSend mechanism. So If I have my main binary, with a `msgSend(obj, "foobar")` and the method `foobar` is implemented in a different binary, than the main binary has no hint where the implementation actually is. It's not even clear in which binary it exists, because it could be part of a superclass of `obj`, which comes from a third library

**Reply by ryanmkurtz** [reply]
After playing with it, this isn't very intuitive.  Maybe the GUI isn't actually the way to go.  Basically if you have the library and function node set up in the Symbol Tree, you can make a function in your EXTERNAL block (for example) and then right click on the function, and do `Function -> Set Thunked Function`, and then for address you'd do `<library name>::<symbol name>`.

I'll get you some code examples from the MachoLoader.

**Reply by ryanmkurtz** [reply]
When I initially make my `ExternalLocation`s, i don't know what library they belong to so I put them in `Library.UNKNOWN` (`<EXTERNAL>`).  You can likely avoid that step.

https://github.com/NationalSecurityAgency/ghidra/blob/af0ea1e7396d922e1cc0f604aa84af8244ba3fd0/Ghidra/Features/Base/src/main/java/ghidra/app/util/opinion/MachoProgramBuilder.java#L627-L632

https://github.com/NationalSecurityAgency/ghidra/blob/af0ea1e7396d922e1cc0f604aa84af8244ba3fd0/Ghidra/Features/Base/src/main/java/ghidra/app/util/opinion/MachoProgramBuilder.java#L854-L866

https://github.com/NationalSecurityAgency/ghidra/blob/af0ea1e7396d922e1cc0f604aa84af8244ba3fd0/Ghidra/Features/Base/src/main/java/ghidra/app/util/opinion/MachoProgramBuilder.java#L1536-L1553
